engine/src/main/java/org/terasology/world/chunks/blockdata/ExtraBlockDataManager.java

```diff
@@ -0,0 +1,305 @@
+               ! Modifier.isStatic(method.getModifiers()) ? "method not static" :
+               argumentTypes.length != 1 ? "arguments list has wrong length" :
+               argumentTypes[0] != Block.class ? "incorrect argument type" :
+               null;
+    }
+    
+    private static class Graph {
+        public String[] verts;
+        public HashMap<String, Set<String>> edges;
+        
+        public Graph(String[] verts, HashMap<String, Set<String>> edges) {
+            this.verts = verts;
+            this.edges = edges;
+        }
+        
+        public Graph(String[] verts) {
+            this.verts = verts;
+            this.edges = new HashMap<>();
+            for (String vert : verts) {
+                edges.put(vert, new HashSet());
+            }
+        }
+        
+        public Graph addEdge(String s0, String s1) {
+            edges.get(s0).add(s1);
+            edges.get(s1).add(s0);
+            return this;
+        }
+        
+        public Graph removeEdge(String s0, String s1) {
+            edges.get(s0).remove(s1);
+            edges.get(s1).remove(s0);
+            return this;
+        }
+        
+        // Creates a new graph containing the complement of the contraction of the complement.
+        public Graph ntract(String s0, String s1) {
+            int v0, v1 = -1;
+            for (int i=0; i<verts.length; i++) {
+                if (verts[i].equals(s0)) {
+                    v0 = i;
+                }
+                if (verts[i].equals(s1)) {
+                    v1 = i;
+                }
+            }
+            String[] newVerts = new String[verts.length-1];
+            System.arraycopy(verts, 0, newVerts, 0, v1);
+            System.arraycopy(verts, v1+1, newVerts, v1, verts.length-v1-1);
+            HashMap<String, Set<String>> newEdges = new HashMap<>();
+            for (String s : verts) {
+                newEdges.put(s, new HashSet<>(edges.get(s)));
+            }
+            newEdges.remove(s1);
+            Set<String> e0 = newEdges.get(s0);
+            Set<String> e1 = edges.get(s1);
+            for (String s2 : verts) {
+                if (e0.contains(s2) && !e1.contains(s2)) {
+                    e0.remove(s2);
+                    newEdges.get(s2).remove(s0);
+                }
+                if (e1.contains(s2)) {
+                    newEdges.get(s2).remove(s1);
+                }
+            }
+            return new Graph(newVerts, newEdges);
+        }
+        
+        public String toString() {
+            String result = "Graph:";
+            for (int i=0; i<verts.length; i++) {
+                result += " (" + verts[i] + " ->";
+                for (String v : edges.get(verts[i])) {
+                    result += " " + v;
+                }
+                result += ")";
+            }
+            return result;
+        }
+    }
+    
+    private static Graph getDisjointnessGraph(Map<String, HashSet<Block>> fields) {
```

Comment:
getDisjointnessGraph is more strongly related to ExtraBlockDataManager than to generic graph things. findCliqueCover is more directly graph-related, and I don't have much opinion either way about which class it would be best in.